File: client/local_game_client.py
import os
import random

from game_interface import GameMessage, Question, TotemQuestion
from solver import Solver


class LocalGameClient:
    def __init__(self, solver: Solver):
        self.solver = solver

    async def run(self):
        print("[Running in local mode]")
        if "SEED" in os.environ:
            random.seed(int(os.environ["SEED"]))
        num_shapes = int(os.environ.get("NUM_SHAPES", "8"))
        SHAPES = ["I", "O", "J", "L", "S", "Z", "T"]
        questions = [TotemQuestion(shape=random.choice(SHAPES)) for _ in range(num_shapes)]
        game_message: GameMessage = GameMessage(
            tick=1, payload=Question(totems=questions)
        )
        self.solver.get_answer(game_message)

File: rust/src/client/local_game_client.rs
use application::{
    game_interface::{GameMessage, Question, Totem, TotemQuestion},
    solver::Solver,
};
use rand::{
    self,
    distributions::{Distribution, Uniform},
};
use std::{env, marker::PhantomData};

pub struct LocalGameClient<S>
where
    S: Solver,
{
    _solver: PhantomData<S>,
}

impl<S> LocalGameClient<S>
where
    S: Solver,
{
    pub fn new() -> Self {
        Self {
            _solver: PhantomData,
        }
    }

    pub async fn run(&self) {
        println!("[Running in local mode]");

        let mut rng = rand::thread_rng();
        let n_totems = match env::var("TOTEMS") {
            Ok(val) => val.parse().unwrap(),
            Err(_) => 8,
        };
        let mut questions: Vec<TotemQuestion> = Vec::new();
        let die = Uniform::from(0..7);
        for _ in 0..n_totems {
            let idx: usize = die.sample(&mut rng);
            let shape: Totem = unsafe { std::mem::transmute(idx) };
            questions.push(TotemQuestion { shape });
        }
        let question = Question { totems: questions };
        let game_message = GameMessage {
            tick: 1,
            payload: question,
        };

        S::solve(&game_message.payload);
    }
}
